### PR TITLE
Attempt to simplify the token cache providers

### PR DESCRIPTION
**Pull request by jmprieur**
DO NOT MERGE yet (until [Expose IsApplicationTokenCache #1448](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1448)) is released in MSAL.NET

Ideally we'd want the implementation of the token cache
providers to be simplified. This is not yet possible because the MSAL Abstract token cache provider contained a boolean telling if the token cache is an app or user token cache (so that the key
can be computed).

Now the TokenCacheEventArgs bring this boolean. 

This PR simplifies the code a lot.

### CODE REVIEW

File: Microsoft.Identity.Web/Microsoft.Identity.Web.csproj [hunk 1]
```diff
@@ -57,6 +57,6 @@
     <PackageReference Include="Microsoft.AspNetCore.Authentication.AzureAD.UI" Version="2.2.0" />
     <PackageReference Include="Microsoft.AspNetCore.Authentication.AzureADB2C.UI" Version="2.2.0" />
     <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="2.2.0" />
-    <PackageReference Include="Microsoft.Identity.Client" Version="4.3.1" />
+    <PackageReference Include="Microsoft.Identity.Client" Version="4.5.1-internal20191022.1isappcache" />
```
Comment (jmprieur) [inline Microsoft.Identity.Web/Microsoft.Identity.Web.csproj]: Experimental. Contains PR https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1448, which not yet merged

**Comment by jmprieur** [comment]
cc: @bgavrilMS @jennyf19 @henrik-me 

@kalyankrishna1 @TiagoBrenck : if you want the MSAL.NET bits that work for this PR, see the artifacts of [this build](https://identitydivision.visualstudio.com/IDDP/_build/results?buildId=416976)

File: Microsoft.Identity.Web/TokenCacheProviders/MsalAbstractTokenCacheProvider.cs [hunk 2]
```diff
@@ -111,13 +102,29 @@ protected virtual Task OnBeforeWriteAsync(TokenCacheNotificationArgs args)
 
         public async Task ClearAsync()
         {
-            await RemoveKeyAsync(CacheKey).ConfigureAwait(false);
+            // This is here a user token cache
```
Comment (jennyf19) [inline Microsoft.Identity.Web/TokenCacheProviders/MsalAbstractTokenCacheProvider.cs]: >This is here a user token cache [](start = 15, length = 31)

nit. "This is a user token cache"

File: Microsoft.Identity.Web/TokenCacheProviders/Sql/SqlTokenCacheProviderExtension.cs [hunk 3]
```diff
@@ -22,37 +20,6 @@ public static class SqlTokenCacheProviderExtension
         public static IServiceCollection AddSqlTokenCaches(
```
Comment (jennyf19) [inline Microsoft.Identity.Web/TokenCacheProviders/Sql/SqlTokenCacheProviderExtension.cs]: nit: Can this be singular? All the others are singular (i think)...ex `AddSqlTokenCache`

**Pull request merged**